azure/flatcar-linux/kubernetes/controllers.tf

```diff
@@ -70,13 +70,14 @@ resource "azurerm_virtual_machine" "controllers" {
     custom_data    = "${element(data.ct_config.controller-ignitions.*.rendered, count.index)}"
   }
 
-  # Azure mandates setting an ssh_key, even though Ignition custom_data handles it too
+  # Azure mandates setting an ssh_key, provide just a single key as the
+  # others are handled with Ignition custom_data
   os_profile_linux_config {
     disable_password_authentication = true
 
     ssh_keys {
       path     = "/home/core/.ssh/authorized_keys"
-      key_data = "${var.ssh_keys}"
+      key_data = "${element(var.ssh_keys, 0)}"
```

Comment:
> Sorry to insist on this, but I'm not sure I understand when this is used. Is this used at all? Or as the first boot happens with ignition, this is overwritten on first boot? Not sure the server state if ignition fails for some reason.

The azure provider requires at least one ssh key when `disable_password_authentication` is `true`. The confusion could  be that there is not any similar requirement in AWS, GCE or Packet and we have to do that to fulfill the requirement.

In this case the key is copied twice to the server, one time by Azure (no idea how) and the other time by ignition.

> I think being very clear about this might help: either use `key_data = "unused see comment"` (or state in the comment that this is set but **never** used in practice) if it is not used **at all in any case**, or try to use all the keys if they are indeed used, isn't that possible?
>

I could have try to use multiple keys, but given that they are already handled by ignition I avoided that. I did something like a minimal-effort approach to fulfill the requirement of having a least one key.

> Also, if this happens to be used in some cases, can't we just assign all the keys? Something like `key_data = "${join("\n",var.ssh_keys)}"` doesn't work?

I did something similar in https://github.com/kinvolk/lokomotive-kubernetes/pull/98, then I found it wasn't working and, as already mentioned, I implemented the fasted solution I found.

If you still think it's a good idea to try to copy all the keys here I can give it a try.

Reply:
> The azure provider requires at least one ssh key when `disable_password_authentication` is `true`. The confusion could be that there is not any similar requirement in AWS, GCE or Packet and we have to do that to fulfill the requirement.
> 
> In this case the key is copied twice to the server, one time by Azure (no idea how) and the other time by ignition.

No, sorry I wasn't clear. I mean what happens for example if we set an invalid ssh key? What are the downsides of doing that?

Or if they are downsides, why not use all the keys? (using a " \n" join as I mentioned https://github.com/kinvolk/lokomotive-kubernetes/pull/116#discussion_r361979875 and @invidian here https://github.com/kinvolk/lokomotive-kubernetes/pull/116#discussion_r362429469)

Reply:
> I did something similar in #98, then I found it wasn't working and, as already mentioned, I implemented the fasted solution I found.

Oh, sorry, just read this. Ok, if it's not working, I'd like to understand if we need this. Like, what is the downside of using an invalid key, something like `key_data = `ssh-rsa ignoreme`?

Reply:
Just for the record. I tried to copy all keys here too but it's not possible to do in Terraform 0.11.

The configuration would look like:

```
os_profile_linux_config {
    disable_password_authentication = true
    ssh_keys {
      path     = "/home/core/.ssh/authorized_keys"
      key_data = "${element(var.ssh_keys, 0)}"
    }
    ssh_keys {
      path     = "/home/core/.ssh/authorized_keys"
      key_data = "${element(var.ssh_keys, 1)}"
    }

  # other ssh_keys blocks...
  }
```

It is not possible to generate all those blocks depending on the length of `var.ssh_keys`.

Reply:
I suggested using join, though, not that :) (https://github.com/kinvolk/lokomotive-kubernetes/pull/116#discussion_r361979875)

Reply:
That was not working either. The only thing  I didn't try was to add a fake ssh key (key_data = "ssh-rsa ignoreme").

Reply:
Sorry, thanks! :)